x/ibc/07-tendermint/misbehaviour.go

```diff
@@ -38,50 +36,34 @@ func CheckMisbehaviourAndUpdateState(
 			"client is already frozen at earlier height %d than misbehaviour height %d", tmClientState.FrozenHeight, misbehaviour.GetHeight())
 	}
 
-	tmConsensusState, ok := consensusState.(*types.ConsensusState)
+	tmConsensusState1, ok := consensusState1.(*types.ConsensusState)
 	if !ok {
-		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "expected type %T, got %T", &types.ConsensusState{}, consensusState)
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "expected type %T, got %T", &types.ConsensusState{}, consensusState1)
+	}
+	tmConsensusState2, ok := consensusState2.(*types.ConsensusState)
+	if !ok {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "expected type %T, got %T", &types.ConsensusState{}, consensusState2)
 	}
 
 	tmEvidence, ok := misbehaviour.(types.Evidence)
```

Comment:
so currently there exists a bug here by not checking `tmEvidence.ValidateBasic()`. Basically we never make an assertion that the message that lead to this callback == `ibctmtypes.SubmitClientMisbehaviour`. The only place `evidence.ValidateBasic` occurs is in this message, but the evidence pkg routes evidence based on what the `Route()` message returns, so I could create a fake message that implements a misbehaviour message, always returns true on validate basic and returns a ibctmtypes.Evidence that has two headers with different heights and even a completely different height for `GetHeight()`. 

The short term solution is to add a `tmEvidence.ValidateBasic`, since we just checked the type of the misbehaviour we can be assured the heights are being properly verified. The long term solution is to add a `ValidateBasic` to the misbehaviour interface and call this in `02-client`

Reply:
this is an existing bug

Reply:
this sort of trickery is something we should write tests for before launch

Reply:
Put in `ValidateBasic` in 02-client for now and added test